### ISSUE TITLE
[BUG] Nullness Checker allows unsound nested generic method calls

### ISSUE THREAD

**Issue by PawelLipski** [typetools/checker-framework#3289]
**Checker version**: 3.4.0

**Affected checkers**: `NullnessChecker`

**The problem**:

Given a generic class:

```java
class GenericClass<T> {}
```

and the following two generic methods:

```java
  <T> GenericClass<T> bar(GenericClass<T> key, T value) {
    return key;
  }

  <T> void qux(GenericClass<T> key) {}
```

in many cases it's possible to circumvent nullness check in nested method calls:

```java
    GenericClass<@NonNull String> x = new GenericClass<>();

    // bar(x, null); // does not compile, okay
    bar((GenericClass<@Nullable String>) x, null); // compiles, okay
    // bar((GenericClass<@NonNull String>) x, null); // does not compile, okay

    // qux(this.<@NonNull String>bar(x, null)); // does not compile, okay
    // qux(this.<@Nullable String>bar(x, null)); // does not compile, okay
    qux(bar(x, null)); // COMPILES BUT SHOULDN'T (?)

    bar((GenericClass<@Nullable String>) x, null); // compiles, okay
    // bar((GenericClass<@NonNull String>) x, null); // does not compile, okay

    qux(bar((GenericClass<@Nullable String>) x, null)); // compiles, okay
    qux(bar((GenericClass<@NonNull String>) x, null)); // COMPILES BUT SHOULDN'T (?)

    this.<@Nullable String>qux(bar(x, null)); // compiles, okay
    this.<@NonNull String>qux(bar(x, null)); // COMPILES BUT SHOULDN'T (?)
```

**Sample repo where the problem can be reproduced**: see the following PR, just run `./gradlew compileJava`: https://github.com/VirtusLab/git-machete-intellij-plugin/pull/258/files

**Comment by wmdietl** [comment]
This is caused by #979 and the work-arounds we currently use to reduce the number of false positives.

You can provide the `-AconservativeUninferredTypeArguments` option to get warnings on the lines where you expect them, plus also false positives that we wouldn't want to have by default.

**Comment by PawelLipski** [comment]
Thanks for quick response. Closing this issue then since it's effectively a subset of another issue.
